certs/manager.go

```diff
@@ -0,0 +1,384 @@
+package certs
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/hex"
+	_ "encoding/json"
+	"encoding/pem"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/pmylund/go-cache"
+)
+
+// StorageHandler is a standard interface to a storage backend,
+// used by AuthorisationManager to read and write key values to the backend
+type StorageHandler interface {
+	GetKey(string) (string, error)
+	SetKey(string, string, int64) error
+	GetKeys(string) []string
+	DeleteKey(string) bool
+	DeleteScanMatch(string) bool
+}
+
+type CertificateManager struct {
+	storage StorageHandler
+	logger  *logrus.Entry
+	cache   *cache.Cache
+	secret  string
+}
+
+func NewCertificateManager(storage StorageHandler, secret string, logger *logrus.Logger) *CertificateManager {
+	if logger == nil {
+		logger = logrus.New()
+	}
+
+	return &CertificateManager{
+		storage: storage,
+		logger:  logger.WithFields(logrus.Fields{"prefix": "cert_storage"}),
+		cache:   cache.New(5*time.Minute, 10*time.Minute),
+		secret:  secret,
+	}
+}
+
+// Extracted from: https://golang.org/src/crypto/tls/tls.go
+//
+// Attempt to parse the given private key DER block. OpenSSL 0.9.8 generates
+// PKCS#1 private keys by default, while OpenSSL 1.0.0 generates PKCS#8 keys.
+// OpenSSL ecparam generates SEC1 EC private keys for ECDSA. We try all three.
+func parsePrivateKey(der []byte) (crypto.PrivateKey, error) {
+	if key, err := x509.ParsePKCS1PrivateKey(der); err == nil {
+		return key, nil
+	}
+	if key, err := x509.ParsePKCS8PrivateKey(der); err == nil {
+		switch key := key.(type) {
+		case *rsa.PrivateKey, *ecdsa.PrivateKey:
+			return key, nil
+		default:
+			return nil, errors.New("tls: found unknown private key type in PKCS#8 wrapping")
+		}
+	}
+
+	if key, err := x509.ParseECPrivateKey(der); err == nil {
+		return key, nil
+	}
+
+	return nil, errors.New("tls: failed to parse private key")
+}
+
+func isSHA256(value string) bool {
+	// 32 SHA256 size * 2 Hex encoding
+	if len(value) != 64 {
+		return false
+	}
+
+	for _, c := range value {
+		if (c < 'a' || c > 'z') && (c < '0' || c > '9') {
+			return false
+		}
+	}
```

Comment:
Replace with `_, err := hex.DecodeString(value); return err == nil`

(perhaps a bit slower, but much simpler and self-explanatory)